Mathlib/Algebra/Ring/Ext.lean

```diff
@@ -0,0 +1,477 @@
+  `(term| (letI := $inst; ($lhs : $type) + ($rhs : $type) : $type))
+local macro:70 lhs:term:70 "*[" type:term ", " inst:term "]" rhs:term:71 : term =>
+  `(term| (letI := $inst; ($lhs : $type) * ($rhs : $type) : $type))
+
+universe u
+
+variable {R : Type u}
+
+/-! ### Distrib -/
+namespace Distrib
+
+@[ext] theorem ext ⦃inst₁ inst₂ : Distrib R⦄
+    (h_add : ∀ x y, x +[R, inst₁] y = x +[R, inst₂] y)
+    (h_mul : ∀ x y, x *[R, inst₁] y = x *[R, inst₂] y) :
+    inst₁ = inst₂ := by
+  -- Split into `add` and `mul` functions and properties.
+  rcases inst₁ with @⟨⟨⟩, ⟨⟩⟩
+  rcases inst₂ with @⟨⟨⟩, ⟨⟩⟩
+  -- Prove equality of parts using function extensionality.
+  congr <;> (apply funext₂; assumption)
+
+theorem ext_iff (inst₁ inst₂ : Distrib R) :
+    inst₁ = inst₂ ↔
+      (∀ x y, x +[R, inst₁] y = x +[R, inst₂] y)
+      ∧ (∀ x y, x *[R, inst₁] y = x *[R, inst₂] y) :=
+  ⟨fun h ↦ by constructor <;> (intros; congr), And.elim (ext · ·)⟩
+
+end Distrib
+
+/-! ### NonUnitalNonAssocSemiring -/
+namespace NonUnitalNonAssocSemiring
+
+@[ext] theorem ext ⦃inst₁ inst₂ : NonUnitalNonAssocSemiring R⦄
+    (h_add : ∀ x y, x +[R, inst₁] y = x +[R, inst₂] y)
+    (h_mul : ∀ x y, x *[R, inst₁] y = x *[R, inst₂] y) :
+    inst₁ = inst₂ := by
+  -- Split into `AddMonoid` instance, `mul` function and properties.
+  rcases inst₁ with @⟨_, ⟨⟩⟩
+  rcases inst₂ with @⟨_, ⟨⟩⟩
+  -- Prove equality of parts using already-proved extensionality lemmas.
+  congr <;> (ext; apply_assumption)
+
+theorem toDistrib_injective : Function.Injective (@toDistrib R) := by
+  intro _ _ h
+  ext x y
+  · exact congrArg (·.toAdd.add x y) h
+  · exact congrArg (·.toMul.mul x y) h
+
+theorem ext_iff (inst₁ inst₂ : NonUnitalNonAssocSemiring R) :
+    inst₁ = inst₂ ↔
+      (∀ x y, x +[R, inst₁] y = x +[R, inst₂] y)
+      ∧ (∀ x y, x *[R, inst₁] y = x *[R, inst₂] y) :=
+  ⟨fun h ↦ by constructor <;> (intros; congr), And.elim (ext · ·)⟩
+
+end NonUnitalNonAssocSemiring
+
+/-! ### NonUnitalSemiring -/
+namespace NonUnitalSemiring
+
+theorem toNonUnitalNonAssocSemiring_injective :
+    Function.Injective (@toNonUnitalNonAssocSemiring R) := by
+  rintro ⟨⟩ ⟨⟩ _; congr
+
+@[ext] theorem ext ⦃inst₁ inst₂ : NonUnitalSemiring R⦄
+    (h_add : ∀ x y, x +[R, inst₁] y = x +[R, inst₂] y)
+    (h_mul : ∀ x y, x *[R, inst₁] y = x *[R, inst₂] y) :
+    inst₁ = inst₂ :=
+  toNonUnitalNonAssocSemiring_injective <|
+    NonUnitalNonAssocSemiring.ext h_add h_mul
+
+theorem ext_iff (inst₁ inst₂ : NonUnitalSemiring R) :
+    inst₁ = inst₂ ↔
+      (∀ x y, x +[R, inst₁] y = x +[R, inst₂] y)
+      ∧ (∀ x y, x *[R, inst₁] y = x *[R, inst₂] y) :=
+  ⟨fun h ↦ by constructor <;> (intros; congr), And.elim (ext · ·)⟩
+
+end NonUnitalSemiring
+
+/-! ### NonAssocSemiring -/
+namespace NonAssocSemiring
+
+/- The best place to prove that the `NatCast` is determined by the other operations is probably in
+an extensionality lemma for `AddMonoidWithOne`, in which case we may as well do the typeclasses
```

Comment:
We should certainly add an ext lemma for `AddMonoidWithOne`

Reply:
Sure; where though?
After the definition of `AddMonoidWithOne` in `Mathlib/Data/Nat/Cast/Defs.lean`, or in a new `Ext.lean` file in the same folder, or in `Mathlib/Algebra/Group/Ext.lean`, or in this file `Mathlib/Algebra/Ring/Ext.lean`? (Or anywhere else?)

Reply:
> or in this file Mathlib/Algebra/Ring/Ext.lean?

I think this is fine. Anywhere is better than in a PR comment!

Reply:
Done in 309b6ed6da.